### ISSUE TITLE
Is there a way to make model.predict_instances run on the gpu?

### ISSUE THREAD

**Issue by DrewScatterday** [mpicbg-csbd/stardist#32]
I passed use_gpu = true into my model when I created the model object inside of my prediction script. When I run ```model.predict_instances``` on my data my CPU jumps to 100% and it seems like my GPU is not being utilized at all. 

Just to clarify I was able to train the model using my GPU so I know it's not an issue with my installation or hardware. I was just hoping I could predict on my data with my GPU to reduce computation time. 

```use_gpu = True and gputools_available()

conf = Config3D (
    use_gpu          = use_gpu,
    n_channel_in     = n_channel,
)
# print(conf)
# vars(conf)

model = StarDist3D(conf, name='stardistAshley', basedir='models')
```

Thanks for any help!

**Comment by maweigert** [comment]
Hi Drew, 

the `use_gpu` flag is only active during training (where it is used to accelerate the data generator). The prediction consists of 2 steps, i) the actual network prediction (which always runs on the GPU if available), and ii) the object post-processing step (which always runs on the CPU with as many cores as it can get). 
So what you observe is indeed the correct behaviour.

Hope that helps, 
Martin

**Comment by DrewScatterday** [comment]
Ahh I see now. Thanks for the quick response! I tried playing around with numba and jit to see if I could force ```predict_instances``` or ```_instances_from_prediction``` to run on the gpu but had no luck as numba is tough to use with a library of objects and functions. 

Do you have any ideas ways I could reduce the computation time of ```predict_instances```? Maybe using multiprocessing/threading with python? 

Its not a huge a deal just was bummed the prediction was taking 3-4 minutes to execute

**Comment by maweigert** [comment]
> Its not a huge a deal just was bummed the prediction was taking 3-4 minutes to execute

The network prediction step and the NMS postprocessing step should roughly (~factor 2) take the same time, unless there is something wrong (e.g. many false very large cell predictions). How large are the images you are predicting on?

> Do you have any ideas on ways I could reduce the computation time of `predict_instances`? Maybe using multiprocessing/threading with python?

The NMS prediction step in fact should use OpenMP to process everything on all the cores of your machine (otherwise it would indeed be very slow) - do you see all CPUs spinning up? If not, the code might not have been compiled with OpenMP enabled, which might be due to using a wrong c/c++ compiler. See [here](https://github.com/mpicbg-csbd/stardist#troubleshooting) for some tips. 

What OS are you on and what is your default c/c++ compiler?

**Comment by DrewScatterday** [comment]
> How large are the images you are predicting on?

The images are (14, 512, 512) (z,y,x)

> What OS are you on and what is your default c/c++ compiler?
I'm working on windows 10 and I'm not 100% sure what c++ compiler but I believe its gcc. I downloaded build tools for visual studio 2019 from [here](https://visualstudio.microsoft.com/downloads/#build-tools-for-visual-studio-2019)

All the CPU's appear to be spinning up as my CPU maxes out to 100% usage on a i7-7770 2.8 GHZ

**Comment by maweigert** [comment]
> All the CPU's appear to be spinning up as my CPU maxes out to 100% usage on a i7-7770 2.8 GHZ

Ok, if you see more than one core going up, than OpenMP should be enabled

> The images are (14, 512, 512) (z,y,x)

For a test image of roughly that size it takes ~10s on my machine (3s prediction, 7s NMS) - how long does it take in your case? 

Additionally could you run the prediction function with verbose output enabled and post the result?

```
y,_ = model.predict_instances(x, n_tiles=(1, 2, 2), nms_kwargs={"verbose":True})
```

**Comment by DrewScatterday** [comment]
Hmmm that's really interesting something seems to be off on my end then. 

I just predicted on data that had a shape of (40, 513, 537) and it was taking 30+ minutes with my CPU maxing out at 100% the entire time. Here's an [image](https://imgur.com/dtrbjM0) of my CPU cores being completely maxed out while predicting. I abandoned the task before it could finish as it was taking a very long time. 

 I reran the prediction with the small data (16, 151, 220) this time using ```nms_kwargs={"verbose":True}``` and it took 6 mins 14 seconds to predict and it still only used the CPU and maxed out the cores. After it ran it gave me the message of:
```
predicting instances with prob_thresh = 0.5 and nms_thresh = 0.4
found 73473 candidates
non-maximum suppression...
NMS took 355.3427 s
keeping 67330/73473 polyhedra
render polygons...
``` 

Whats also weird is my GPU [memory](https://imgur.com/rHeEs6D) almost maxes out while the CPU is running. It's almost like the data gets loaded into the GPU but the GPU doesn't do any computations. 

Its also interesting because I was able to get the GPU to work for training the neural network. 

For reference I'm on windows 10 64bit. I have cuDNN 7.6.5, NVIDIA GPU Computing Toolkit 10.0.130, I manually installed the latest version of pyopencl instead of using pip (because windows is a nightmare) and pulled it from their [github](https://github.com/inducer/pyopencl), python 3.7.3 64bit, c++ compiler from build tools for visual studio 2019 from [here](https://visualstudio.microsoft.com/downloads/#build-tools-for-visual-studio-2019), tensorflow-gpu==1.14, keras==2.2.5, and gputools installed via pip

**Comment by maweigert** [comment]
> After it ran it gave me the message of:

Aha. 73473 object candidates are suspiciously many for that small volume. 
Some thoughts: 

1) Was the input properly normalized? 
2) Have a look at the object probability map directly, that can be obtained via

```python
prob, _ = model.predict(img)
```
and see whether it is concentrated at the object centers

**Comment by DrewScatterday** [comment]
> Was the input properly normalized?

So to normalize I used
 ```img = normalize(X[0], 1,99.8, axis=axis_norm)```  
```labels, details = model.predict_instances(img)```
where ```axis_norm = (0,1,2)```

> Have a look at the object probability map directly, that can be obtained via

Here's what the probability map looked like on one slice of the stack. I'm assuming the probability map should be a smaller point located / / near the center rather than the whole cell highlighted. So maybe thats where the error is occurred? Seems like its predicted a ton of object candidates rather than the 5-6 actual candidates in our image. 

![slice](https://user-images.githubusercontent.com/28267620/72114784-aa6e3a80-3312-11ea-9359-d90f0aa36e69.png)

**Comment by maweigert** [comment]
> Here's what the probability map looked like

Yes, that looks weird and might be the reason for the overabundance of candidates. Do you get the same problems when predicting on a stack of the training set? And are your loss curves (e.g. prob_loss) looking reasonable?

**Comment by DrewScatterday** [comment]
I’ll check the training stack when I get to my computer. I’ll take a look at that probability map and see what it looks like. 

Additionally when you predicted on the (14, 512, 512) image and it only took your machine 10s to predict, how much data was in your training set and validation set when training that neural network? 

I could see this issue on my side potentially being too few data examples given to it during training m causing it to predict lots of candidates causing the computation taking a long time.

**Comment by DrewScatterday** [comment]
Any ideas on how to fix this @maweigert ? 

I just retrained another model on the GPU with 200 epochs using 7 training examples and 1 for validation. My training and validation data were size (50,409,409) -> (Z,Y,X). Training took about an hour with loss getting down to 0.2973. 

I predicted on one of the training data of size (50,409,409) and the prediction takes 10-15 minutes with CPU cores maxing out the whole time. When I ran it with ```nms_kwargs={"verbose":True}``` I received: 
```
predicting instances with prob_thresh = 0.5 and nms_thresh = 0.4
found 15128 candidates
non-maximum suppression...
NMS took 892.2884 s
keeping 127/15128 polyhedra
render polygons...
```

So it seems that is still finding TONS of polygons causing the prediction time to take a long time. I thought by adding more training data that would help reduce prediction computation time but I've had no luck in doing so. 

This is what my prob loss function looks like so it seems that this could be the culprit. My prob loss function on a different model with different data performed way better so this raises suspicion for me. 
![prob](https://user-images.githubusercontent.com/28267620/72195296-9c3a2000-33df-11ea-9ebb-9f7a2f2dc7dd.PNG)

Some ideas I had is maybe my train_patch_size during training is too small? Maybe I'm not normalizing before the model prediction.

Would appreciate any ideas on things to try to get the prediction run faster, thanks for any help!!

**Comment by maweigert** [comment]
Mhhh, hard to debug without seeing the data. 
If you could provide i) one of your trainings stacks, and ii) the model folder, I can have a quick look into it.

**Comment by DrewScatterday** [comment]
This has the training image.tif and the mask for it at mask.tif. Then the model folder is in there as well. Appreciate any help! 

[githubdebug.zip](https://github.com/mpicbg-csbd/stardist/files/4048892/githubdebug.zip)

**Comment by uschmidt83** [comment]
Hi Drew,

I think the issue here is that your StarDist model isn't very well trained yet (I'm guessing, since I haven't seen the ground truth segmentation masks). The predicted shapes are rather jagged, which causes the NMS postprocessing step to run much less efficient.

![image](https://user-images.githubusercontent.com/231996/72206825-96e9dd80-3492-11ea-9731-fc2cb71b83fc.png)

Verbose NMS messages:
```
NMS: Function calls:
NMS: ~ bbox+out:  1140258
NMS: ~ inner:       19411
NMS: ~ kernel:      19409
NMS: ~ convex:      14541
NMS: ~ render:      13074
NMS: Excluded intersection:
NMS: + pretest:   1120847
NMS: + convex:       1467
NMS: Suppressed polyhedra:
NMS: # inner:           2 / 15128  (0.01 %)
NMS: # kernel:       4868 / 15128  (32.18 %)
NMS: # render:      10131 / 15128  (66.97 %)
NMS: # total:       15001 / 15128  (99.16 %)
```

**Comment by DrewScatterday** [comment]
Hmm interesting. I gathered more labeled data examples so I'm going to retrain with 25 training examples with 400 epochs and a different sized ```train_patch_size``` and see if that yields better results

**Comment by maweigert** [comment]
> This has the training image.tif and the mask for it at mask.tif. Then the model folder is in there as well. Appreciate any help!

Some comments:
1) Yes, the patchsize you used (`16,64,64`) is rather small, especially as some of your cells are pretty big. I would therefore recommend using larger patches, e.g. `train_patch_size = (64,196,196)` and `train_batch_size=1` (such that it still fits into GPU memory).
2) Your annotated GT images(`mask.tif`) seems to have some issues:
<img width="400" alt="Screenshot 2020-01-11 at 21 38 02" src="https://user-images.githubusercontent.com/11042162/72210374-ad0b9400-34ba-11ea-80a5-7c11b235603f.png">
There appear many "annotation holes" within the cell body which would massively disturb the training. E.g. in the large green cell in the middle (which should in fact be two?). I would therefore recommend to fix the GT too.

Hope that helps!

**Comment by DrewScatterday** [comment]
Ahhh I see. I was trying to use a larger ```train_patch_size``` but I keep running out GPU memory but that makes so much more sense now with the ```train_batch_size```. Thanks for the help! I'll run some training overnight and take a look it at tomorrow to see how it does.

**Comment by DrewScatterday** [comment]
Ultimately I figured out my issue was the the nonmax suppression and object probability threshold wasn't be optimized which was was causing my prediction to take an extremely long time. Thanks for the help guys!
